### ISSUE TITLE
Undesired error report when schema is changed (not null column is dropped) in rc or for-update-read.

### ISSUE THREAD

**Issue by SpadeA-Tang** [pingcap/tidb#34978]
## Bug Report

This is another variant of #21498.

### 1. Minimal reproduce step (Required)

```
create table t (id int primary key, c int not null)
insert into t values (1, 1), (2, 2)
```
session 1
```
begin
```
session 2
```
alter table t drop column c
insert into t values
```
session 1
```
select * from t for update  // error
```

In `READ-COMMITTED`, the error is the same.

### 3. What did you see instead (Required)
`ERROR 1105 (HY000): [components/tidb_query_executors/src/table_scan_executor.rs:411]: Data is corrupted, missing data for NOT NULL column (offset = 1)`

**Comment by cfzjywxk** [comment]
@SpadeA-Tang 
Better to link this issue to the existing schema issues.

**Comment by SpadeA-Tang** [comment]
> @SpadeA-Tang Better to link this issue to the existing schema issues.

Okay，done.

**Comment by cfzjywxk** [comment]
Fixed on master branch a6d194b3c71d287c8ca403cc962e1ede3c73b83e after the MDL is introduced.
